lib/assets/javascripts/turbograft/turbolinks.coffee

```diff
@@ -121,23 +121,22 @@ class window.Turbolinks
     for k,v of options.headers
       xhr.setRequestHeader k, v
 
-    xhr.onload = ->
-      if xhr.status >= 500
-        Turbolinks.fullPageNavigate(url)
-      else
-        Turbolinks.loadPage(url, xhr, options)
-      xhr = null
-
-    xhr.onerror = ->
-      # Workaround for sinon xhr.abort()
-      if xhr.statusText == "abort"
+    new Promise (resolve) => 
+      xhr.onload = ->
+        if xhr.status >= 500
+          Turbolinks.fullPageNavigate(url)
+        else
+          resolve(Turbolinks.loadPage(url, xhr, options))
         xhr = null
-        return
-      Turbolinks.fullPageNavigate(url)
 
-    xhr.send()
+      xhr.onerror = ->
+        # Workaround for sinon xhr.abort()
+        if xhr.statusText == "abort"
+          xhr = null
+          return
```

Comment:
Do you need to resolve with a copy of the XHR (here, and above)? IIRC, rejecting here would cause the promise `.fail` methods to be called, but people might want to treat a real failure (HTTP 500+) differently from this abort (e.g., don't flash an error message or whatever they're doing).  People might want to inspect the XHR (I've written some code that checks the status for abort and behaves differently) for `xhr.statusText == 'abort'` and react accordingly.

This path would occur if you were to call `Page.refresh` twice in quick succession right (and cared to bind to the promise)?